### PR TITLE
Add RBAC decorator and plugin

### PR DESCRIPTION
**Pull request by davidolrik**
Role based authentication for use in plugins.

Included in this pull request is two decorators: `rbac_require_any_role` and `rbac_require_all_role`, and a "admin" plugin for assigning roles dynamically, and for being notified of people trying to use commands they aren't allowed to.

```python
# User must have any one of the listed roles
@respond_to(regex=r'^my cool command$')
@rbac_require_any_role(['root', 'admin'])
def my_cool_comand(self, msg):
    pass

# User must have all of the listed roles
@respond_to(regex=r'^my other command$')
@rbac_require_all_roles(['role1', 'role2'])
def my_other_comand(self, msg):
    pass
```

Still needs docs and better error handling for edge cases.

### CODE REVIEW

**Comment by DandyDev** [comment]
This is super useful! Needs docs indeed.

In my opinion, we could get rid of the `rbac` prefixes in the decorator. It's clear from the rest of the decorator names what it's doing.

**Comment by DandyDev** [comment]
@davidolrik I'd still love to include this, but it still needs documentation. Are you planning to provide that?

**Comment by davidolrik** [comment]
Definitely!

**Comment by davidolrik** [comment]
@DandyDev: I've added docs - let me know if it is enough, and if it is placed correctly.

**Comment by davidolrik** [comment]
I've also rebased it with `master` so it should be easy to merge.

**Comment by davidolrik** [comment]
Rebased with master.

**Comment by davidolrik** [comment]
Rebased with master.

**Comment by davidolrik** [comment]
@DandyDev Are there anything more needed for this to be merged?

File: docs/plugins/interacting.rst [hunk 1]
```diff
@@ -155,6 +155,58 @@ There is one caveat:
...
+
+You can get the *member ID* from the slack profile, by clicking *more* and
+selecting *Copy member ID*.
+
+If you wish to share the powers of root you can enable the RBAC admin plugin
+:py:class:`~machine.plugins.builtin.admin.RBACPlugin` and grant the *admin* role
+to users you trust.
+
+The RBAC plugin provides you with three new commands that lets you lookup, grant
+and revoke roles to users: *@superbot who has role admin*, *@superbot grant
```
Comment (DandyDev) [inline docs/plugins/interacting.rst:189]: The word `role` is missing here?

**Comment by DandyDev** [comment]
Sorry for the super late response. I was switching jobs and have a new puppy that takes a lot of my spare time. This looks great and there's only one small change in the docs that is required before I can merge.

File: docs/plugins/interacting.rst (hunk 1 above)
Comment (davidolrik) [reply docs/plugins/interacting.rst]: Nice catch! - Fixed in 1aa1389

**Comment by davidolrik** [comment]
No worries =) - I have a Golden my self, which kind of puppy?

**Comment by DandyDev** [comment]
image

A golden as well! <3

**Comment by davidolrik** [comment]
A9B80BFF-74FD-4B40-8386-83C1A86C06B7_1_105_c

**Comment by DandyDev** [comment]
What a lovely doggo!

This PR has turned out to be the most wholesome PR in my open source career so far ;)

**Comment by DandyDev** [comment]
Merged!

**Comment by davidolrik** [comment]
Lets make more PR's like this then 😁

**Pull request merged**